src/dbally/prompt/aggregation.py

```diff
@@ -0,0 +1,122 @@
+    """
+    Validates LLM response to IQL
+
+    Args:
+        llm_response: LLM response
+
+    Returns:
+        A string containing aggregations.
+
+    Raises:
+        UnsupportedAggregationError: When IQL generator is unable to construct a query
+        with given aggregation.
+    """
+    if "unsupported query" in llm_response.lower():
+        raise UnsupportedAggregationError
+    return llm_response
+
+
+class AggregationPromptFormat(PromptFormat):
+    """
+    Aggregation prompt format, providing a question and aggregation to be used in the conversation.
+    """
+
+    def __init__(
+        self,
+        question: str,
+        aggregations: List[ExposedFunction] = None,
+    ) -> None:
+        super().__init__()
+        self.question = question
+        self.aggregations = "\n".join([str(aggregation) for aggregation in aggregations]) if aggregations else []
+
+
+class AggregationFormatter:
+    """
+    Class used to manage choice and formatting of aggregation based on natural language question.
+    """
+
+    def __init__(self, llm: LLM, prompt_template: Optional[PromptTemplate[AggregationPromptFormat]] = None) -> None:
+        """
+        Constructs a new AggregationFormatter instance.
+
+        Args:
+            llm: LLM used to generate IQL
+            prompt_template: If not provided by the users is set to `AGGREGATION_GENERATION_TEMPLATE`
+        """
+        self._llm = llm
+        self._prompt_template = prompt_template or AGGREGATION_GENERATION_TEMPLATE
+
+    async def format_to_query_object(
+        self,
+        question: str,
+        event_tracker: EventTracker,
+        aggregations: List[ExposedFunction] = None,
+        llm_options: Optional[LLMOptions] = None,
+    ) -> IQLQuery:
+        """
+        Generates IQL in text form using LLM.
+
+        Args:
+            question: User question.
+            event_tracker: Event store used to audit the generation process.
+            aggregations: List of aggregations exposed by the view.
+            llm_options: Options to use for the LLM client.
+
+        Returns:
+            Generated aggregation query.
+        """
+        prompt_format = AggregationPromptFormat(
+            question=question,
+            aggregations=aggregations,
+        )
+
+        formatted_prompt = self._prompt_template.format_prompt(prompt_format)
+
+        response = await self._llm.generate_text(
+            prompt=formatted_prompt,
+            event_tracker=event_tracker,
+            options=llm_options,
+        )
+        # TODO: Move response parsing to llm generate_text method
+        agg = formatted_prompt.response_parser(response)
+        # TODO: Move IQL query parsing to prompt response parser
+        return await IQLQuery.parse(
+            source=agg,
+            allowed_functions=aggregations or [],
+            event_tracker=event_tracker,
+        )
+
+
+AGGREGATION_GENERATION_TEMPLATE = PromptTemplate[AggregationPromptFormat](
```

Comment:
I think we should first ask the LLM whether to use aggregation at all before asking it to generate it. In this implementation, there is no way to skip this step, which introduces hallucinated answers or `UnsupportedAggregationError` for filter-only questions.